### DISCUSSION TITLE
Globally available components

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#2371] [Q&A]
What is the best way to create a component that is globally available?

For example, I've created custom controllers for file accessing, so that I can have the Url that I want to. Therefore, now I require a custom File Manager that, instead of generating the usual url (api/file/download/id), generates my custom url. This component should be globally available for all modules, just like the File Manager in Oqtane.Client/Modules/Controls.

How can I achieve this without modifying base Oqtane solution?

**Comment by sbwalker** [comment]
The FileManager component cannot be replaced with another version without modifying the core framework. The simplest solution would be to replace the FileManager.razor source with your own implementation and then rebuild the framework. In the future, the plan is for the framework to support multiple "folder providers" - an abstraction which will cleanly separate the storage of files from the API. Today there are 2 definitions for folder types - Private (using api/file/download/id) and Public (using standard wwwroot relative paths) - however the folder types are static (in the future they will be extensible). The core FileManager component will be able to accommodate any folder type.

**Reply by mdmontesinos** [reply]
I should explain better what I meant. I'm not changing folder provider, I'm just changing the way I can access the files.

To do that, I created a new controller like this:

```c#
[Route("[controller]")]
    public class PdfController : Controller
    {
        private readonly IWebHostEnvironment _environment;
        private readonly IFileRepository _files;
        private readonly IFolderRepository _folders;
        private readonly IUserPermissions _userPermissions;
        private readonly ILogManager _logger;
        private readonly Alias _alias;
        private readonly string _pdfFolderPath = "Pdf";

        public PdfController(IWebHostEnvironment environment, IFileRepository files, IFolderRepository folders, IUserPermissions userPermissions, ILogManager logger, ITenantManager tenantManager)
        {
            _environment = environment;
            _files = files;
            _folders = folders;
            _userPermissions = userPermissions;
            _logger = logger;
            _alias = tenantManager.GetAlias();
        }

        // GET <controller>/file.pdf
        [HttpGet("{pdfName}")]
        public IActionResult DownloadPdf(string pdfName)
        {
            string errorPath;

            string path = Utilities.PathCombine(_pdfFolderPath, Path.DirectorySeparatorChar.ToString());

            var pdfFolder = _folders.GetFolder(_alias.SiteId, path);

            if (pdfFolder == null)
            {
                _logger.Log(LogLevel.Error, this, LogFunction.Read, $"Pdf Folder Does Not Exist: {_pdfFolderPath}");
                HttpContext.Response.StatusCode = (int)HttpStatusCode.NotFound;

                errorPath = Path.Combine(GetFolderPath("wwwroot\\images"), "error.png");
                return System.IO.File.Exists(errorPath) ? PhysicalFile(errorPath, MimeUtilities.GetMimeType(errorPath)) : null;
            }

            var files = _files.GetFiles(pdfFolder.FolderId);

            var file = files.Where(x => x.Name.Equals(pdfName) && x.Extension.Equals("pdf")).FirstOrDefault();

            if (file != null)
            {
                return Download(file.FileId, false);
            }

            _logger.Log(LogLevel.Error, this, LogFunction.Read, $"File Does Not Exist: {pdfName}");
            HttpContext.Response.StatusCode = (int)HttpStatusCode.NotFound;

            errorPath = Path.Combine(GetFolderPath("wwwroot\\images"), "error.png");
            return System.IO.File.Exists(errorPath) ? PhysicalFile(errorPath, MimeUtilities.GetMimeType(errorPath)) : null;
        }

        private IActionResult Download(int id, bool asAttachment)
        {
            //It's the same method found in FileController
        }
        private string GetFolderPath(string folder)
        {
            //It's the same method found in FileController
        }
    }
```
With it, I just add a new Url to access a Pdf file, that must be stored in a folder called "Pdf", without changing the default one:

- Content/Tenants/1/Sites/1/Pdf/name.pdf (api/file/download/id, if it's private)
- Pdf/name.pdf

Both return exactly the same page.

Now, going back to my original question, I just need to design a "File Manager" ("PdfManager") that only has access to "Pdf" folder, and upon selecting the file, returns, for example, an anchor element with href being "Pdf/name.pdf" instead of "Content/Tenants/1/Sites/1/Pdf/name.pdf". I won't modify the original one, I just need this new one to be globally available to all modules. The same would apply for images, where upon inserting them, the src for the img element would be the new url ("images/name.jpg").

**Reply by sbwalker** [reply]
@mdmontesinos this approach is not compatible with the multi-tenant capabilities of the framework. The reason why the framework uses the folder hierarchy Content/Tenants/1/Sites/1 is to ensure that files are physically separated for each tenant/site. Your solution will only support a single site for a single tenant - which may be fine for your current use case but is a limitation if you ever decide to use the multi-tenant aspects of the framework. But getting back to your original question, there is no way for the FileManager component to integrate with this Controller you created other than modifying the core code.

**Reply by mdmontesinos** [reply]
Perhaps there's something I'm not fully understanding, but I believe it does still support multi-tenant capabilities.

The files are still being stored with the folder hierarchy "Content/Tenants/1/Sites/1", and the only thing I'm doing is creating a new route to access them. However, the new controller does check the current tenant via:
```
var pdfFolder = _folders.GetFolder(_alias.SiteId, path);
```
So if the folder does not exist for the current tenant, it would return an error, therefore preserving multi-tenant aspects. Also the 'Download' method is the same as you would use for private files (api/file/download) in File Manager.

Am I getting something wrong?

As for my original question, and leaving aside the file manager, is there no way then to create a component that is available to every external module without modifying core code or adding a reference in all of them?

**Reply by sbwalker** [reply]
@mdmontesinos apologies... I think I came to the wrong conclusion when I reviewed your code... it should support multi-tenancy based on your explanation.

In regards to creating "global" components, it is absolutely possible. You would need to use a dynamic approach for instantiating them (which is exactly the approach which Oqtane itself uses to instantiate external modules). Basically in your consuming component you can use the DynamicComponent:

```
<DynamicComponent Type="@ComponentType" Parameters="@Parameters"></DynamicComponent>

@code {
	Type ComponentType = Type.GetType("Your Fully Qualified Type Name");
	private IDictionary<string, object> Parameters { get; set; }

	protected override void OnInitialized()
	{
		Parameters = new Dictionary<string, object>();
		Parameters.Add(new KeyValuePair<string, object>("Parameter Name", "Parameter Value"));
	}
}
```
Note that a fully qualified type name contains both the type and the assembly it resides in ie. "YourTypeName, YourAssemblyName". When using this loosely coupled approach you do not need to create a static reference between your consuming component and the global component - it is resolved at runtime.

**Reply by mdmontesinos** [reply]
Great, I'll try it as soon as I can!
